components/script/dom/document.rs

```diff
@@ -829,27 +831,37 @@ impl Document {
             // Update the focus state for all elements in the focus chain.
             // https://html.spec.whatwg.org/multipage/#focus-chain
             if focus_type == FocusType::Element {
-                self.send_to_constellation(ScriptMsg::Focus);
+                self.window().send_to_constellation(ScriptMsg::Focus);
             }
 
             // Notify the embedder to display an input method.
             if let Some(kind) = elem.input_method_type() {
-                self.send_to_constellation(ScriptMsg::ShowIME(kind));
+                let top_level_browsing_context_id = self.window().top_level_browsing_context_id();
+                self.send_to_embedder(EmbedderMsg::ShowIME(top_level_browsing_context_id, kind));
             }
         }
     }
 
     /// Handles any updates when the document's title has changed.
     pub fn title_changed(&self) {
         if self.browsing_context().is_some() {
-            self.send_title_to_constellation();
+            self.send_title_to_embedder();
         }
     }
 
     /// Sends this document's title to the constellation.
-    pub fn send_title_to_constellation(&self) {
-        let title = Some(String::from(self.Title()));
-        self.send_to_constellation(ScriptMsg::SetTitle(title));
+    pub fn send_title_to_embedder(&self) {
+        let window = self.window();
+        if window.is_top_level() {
+            let title = Some(String::from(self.Title()));
+            let top_level_browsing_context_id = window.top_level_browsing_context_id();
```

Comment:
The same way that `send_to_constellation` attaches the browsing context id to each message, could we make it so that any message from pipeline to embedder automatically comes with the browsing context id?

Maybe an embedder message should the message + browser_id?

Reply:
Perhaps this pattern really was better: https://github.com/servo/servo/pull/20707/files#diff-55c92a6a5ba7654ce45fe6fc6c63740fL1183

Reply:
Well, given https://github.com/servo/servo/pull/20707#issuecomment-385470619 we do really want to move away from having one script msg for each emebedder one, so we would need to find another way to add the top_level_id(and it needs to be done at the point where the msg is first created I think)